### DISCUSSION TITLE
Allow nested relational data in button-links interface

### DISCUSSION THREAD

**Discussion by TheAzack9** [directus/directus#11255] [Feature Requests]
Note: I could not find another issue while quickly browsing, but someone might have had similar issues.

It seems that when editing a collection it does a shallow-get for the values and never hydrates the entire object. This will cause e.g the interface presentation-links to not be able to render data from relations.

If you print out the data you get from the "values" provided e.g here: https://github.com/directus/directus/blob/main/app/src/interfaces/presentation-links/presentation-links.vue#L44
it seems that the relations never gets hydrated and just return an array with ids. 

```json
{
  "id": 1,
  "status": "published",
  "sort": null,
  "user_updated": "98188120-7322-4714-93a3-7ac0262a3ddb",
  "date_updated": "2022-01-14T18:48:20",
  "test": "test",
  "metadata": 1,
  "relation": [2, 3]
}
```

When then trying to retreive said data using the preview-links url template then it will just be empty and not open the page i want.
(In my example i tried `https://google.com/test/{{Slug in metadata}}`)

The interesting case here is that if you edit a value in the relation, then it hydrates with that change and suddenly you get expected data. However, after saving it goes back to the old dataset.

```json
{
  "id": 1,
  "status": "published",
  "sort": null,
  "user_updated": "98188120-7322-4714-93a3-7ac0262a3ddb",
  "date_updated": "2022-01-14T18:48:20",
  "test": "test",
  "metadata": { "id": 1, "slug": "So now it works?" },
  "relation": [2, 3]
}
```

I have also noticed that this also causes an error with translations in relations where if you edit the second language in a translations then it simply "deletes" the first one because the data ends up looking like this:

```json
{
  "id": 1,
  "status": "published",
  "sort": null,
  "user_updated": "98188120-7322-4714-93a3-7ac0262a3ddb",
  "date_updated": "2022-01-16T12:01:54",
  "test": "test",
  "metadata": 1,
  "relation": [
    {
      "id": 2,
      "test_relation_id": {
        "id": 1,
        "translations": [
          {
            "id": 1,
            "test_relation_id": 1,
            "languages_code": "de-DE",
            "text": "shit it removed the other one because it isn't hydrated :c"
          },
          2
        ]
      }
    },
    3
  ]
}
```

This is quite an annoying bug and has sadly caused me quite a bit of grief of lost translations for no apparent reason to me 😅

And without having the slightest clue of vue or how the app is structured are my suspicions directed to this for not fetching the whole item, but just the shallow one: https://github.com/directus/directus/blob/main/app/src/composables/use-item/use-item.ts#L100, but idk  

### To Reproduce

1. Create a collection named Metadata with a string field configured to be a slug. 
2. Create another collection e.g "Page" with a Many to one relation to Metadata.
3. In page, create a Button links interface with a link named "Preview" and url `https://google.com/test/{{slug}}` 
4. Create an item of Page and add a metadata with data in the slug field.
5. Save and exit the collection (This is important, check explaination above of why :)  If you press it while editing it seems to work fine)
6. Go back to the item and press the "Preview" button and it will open with no slug.

**Comment by joselcvarela** [comment]
Hey @TheAzack9 
For the Preview Link value try to put this as raw value, because in my end it seems to be working as expected:
`https://google.com/test/{{metadata.slug}}`


https://user-images.githubusercontent.com/14039341/149781643-df627ffa-6820-4eab-b4e9-6ba0b4d3a5cb.mov

**Comment by TheAzack9** [comment]
It happens inside the item after it is created. Templating the url itself is fine, but when i try to actually press the preview it fails due to the app not having the relational values 😓 
I can send over an sqlite database later today with data needed to reproduce potentially :)

**Comment by joselcvarela** [comment]
But if the button link is dependent on the relational items and if the item does not have any relational items, what you would expect to happen?
Yeah it will be useful for us 👍

**Comment by TheAzack9** [comment]
Sorry for taking a bit on uploading, but here is a barebones example made with the latest directus version (using SQLite)
[example.zip](https://github.com/directus/directus/files/7886675/example.zip)
 
Video highlighting the issue:

https://user-images.githubusercontent.com/2140448/149882304-3d84a4c4-2f20-4b1e-8c84-9ea50fe79635.mp4

**Comment by TheAzack9** [comment]
Here is another video highlighting what i mean when i said it works while editing because the edits are populated in the values in the form :)

https://user-images.githubusercontent.com/2140448/149882555-67d74570-80ae-4b45-a7fa-d9f04e8ef3c9.mp4

**Comment by TheAzack9** [comment]
And for good measure, here is a video showing the translations disappearing, which i suspect is for the same reason :) (might possibly be an entierly different issue tho)

https://user-images.githubusercontent.com/2140448/149882899-0e15ca5a-1019-47df-80fa-03fbc12748c3.mp4

**Comment by joselcvarela** [comment]
The issue here is the since the value interpolated comes from a relationship (nested level) and since this interface does not fetch nested values then it's not possible to get the right value.
For example, this is the main item:
```
{ id: 1, metadata: 1}
```

When trying to render `{{metadata.slug}}` it will not work because the current value is 1.

**Comment by joselcvarela** [comment]
@TheAzack9 about the translations you might want to create another issue since they are not related 👍

**Comment by rijkvanzanten** [comment]
Moved this into a feature request, as it's technically working as intended 🙂 

Wanna try your hand at a PR to have the button-links interface fetch it's own data nested as per the configured data structure @TheAzack9?

**Reply by TheAzack9** [reply]
I would gladly! However i'm not quite sure how it would be done in the best way 😅 So i'll gladly make some if i get some pointers :) 
With the limited knowledge of the codebase is there mostly two solutions that i can think of and idk if i'm to fond of either 😅 

1. fetch a deeply nested version of the collection item and use it as a fallback in the case that the data doesn't exist in the injected values in [presentation-links.vue](https://github.com/directus/directus/blob/main/app/src/interfaces/presentation-links/presentation-links.vue#L44). (Worth noting that this would not fix the translation issue, but it is isolated and wouldn't break anything else. However it would significantly slow down presentation-links)
2. Change https://github.com/directus/directus/blob/main/app/src/composables/use-item/use-item.ts#L100 to fetch a deeply nested version of the item. (I could see this cause a lot of issues if it wasn't planned to be this way from the start, which it doesn't seem to have been. It would slow down the initial load of editing a collection quite a bit, but it would solve both button links and the translation bug in the long run. I don't really like this solution particularly much tho :) )

There are probably other, better solutions that i'm missing tho 😅

**Comment by TheAzack9** [comment]
I created a separate issue for the translation bug as i assume that doesn't work as intended 😅 https://github.com/directus/directus/issues/11259

**Comment by azrikahar** [comment]
Related discussions: #5947

---

Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by RANSoftSS** [comment]
Hi, if i want to reopen that PR shoult i create a new one?

**Reply by rijkvanzanten** [reply]
Sure! A new feature request can be created here: https://github.com/directus/directus/discussions/new?category=feature-requests
